File: EDD2026/Arbol/arbol.py
class ArbolBin:
    def __init__(self, izq=None, elem=None, der=None):
        """Constructor nodo: recibe subárbol izquierdo, elemento, subárbol derecho"""
        self.izq = izq
        self.elem = elem
        self.der = der

    @classmethod
    def nulo(cls):
        """Constructor del árbol vacío"""
        return cls(None, None, None)

    def es_nulo(self):
        return self.elem is None

    def es_hoja(self):
        if self.es_nulo():
            return False
        return self.izq is None and self.der is None

    def es_ult(self):
        """Es último? (al menos un hijo es nulo)"""
        if self.es_nulo():
            return False
        return self.izq is None or self.der is None

    def num_nodos(self):
        if self.es_nulo():
            return 0
        izquierda = self.izq.num_nodos() if self.izq else 0
        derecha = self.der.num_nodos() if self.der else 0
        return 1 + izquierda + derecha

    def altura(self):
        if self.es_nulo():
            return 0
        izquierda = self.izq.altura() if self.izq else 0
        derecha = self.der.altura() if self.der else 0
        return 1 + max(izquierda, derecha)

    def camino_min(self):
        if self.es_nulo():
            return 0
        izquierda = self.izq.camino_min() if self.izq else 0
        derecha = self.der.camino_min() if self.der else 0
        return 1 + min(izquierda, derecha)

    def es_completo(self):
        return self.altura() == self.camino_min()

    def num_hojas(self):
        if self.es_nulo():
            return 0
        if self.es_hoja():
            return 1
        izquierda = self.izq.num_hojas() if self.izq else 0
        derecha = self.der.num_hojas() if self.der else 0
        return izquierda + derecha

    def preorden(self):
        """Recorrido en preorden (mostrando nulos)"""
        if self.es_nulo():
            return ["null"]
        resultado = [self.elem]
        if self.izq:
            resultado += self.izq.preorden()
        else:
            resultado += ["null"]
        if self.der:
            resultado += self.der.preorden()
        else:
            resultado += ["null"]
        return resultado


# Construcción del árbol usando la expresión algebraica:
# nodo(nodo(nodo(nulo,2,nulo),8,nulo),3,nodo(nodo(nulo,1,nulo),5,nodo(nodo(nulo,9,nulo),6,nulo)))

nulo = ArbolBin.nulo()

# Construcción desde abajo hacia arriba
nodo2 = ArbolBin(nulo, 2, nulo)
nodo8 = ArbolBin(nodo2, 8, nulo)

nodo9 = ArbolBin(nulo, 9, nulo)
nodo1 = ArbolBin(nulo, 1, nulo)
nodo6 = ArbolBin(nodo9, 6, nulo)
nodo5 = ArbolBin(nodo1, 5, nodo6)

arbol = ArbolBin(nodo8, 3, nodo5)

# Demostración
print("=== ÁRBOL BINARIO ===")
print("Preorden:", arbol.preorden())
print("¿Es nulo?", arbol.es_nulo())
print("Número de nodos:", arbol.num_nodos())
print("Número de hojas:", arbol.num_hojas())
print("Altura:", arbol.altura())
print("Camino mínimo:", arbol.camino_min())
print("¿Es completo?", arbol.es_completo())
print("¿Es último? (raíz):", arbol.es_ult())

File: EDD2026/Arbol/arbol.cpp
#include <iostream>
#include <algorithm>
using namespace std;

template <class Elem>
class ArbolBin {
private:
    struct Nodo {
        Elem elem;
        Nodo* izq;
        Nodo* der;
        Nodo(Nodo* i, Elem e, Nodo* d) : izq(i), elem(e), der(d) {}
    };
    Nodo* raiz;

public:
    // Constructor nulo
    ArbolBin() : raiz(nullptr) {}

    // Constructor nodo
    ArbolBin(ArbolBin* i, Elem e, ArbolBin* d) {
        raiz = new Nodo(i->raiz, e, d->raiz);
    }

    // Constructor interno para nodos hoja
    ArbolBin(Elem e) {
        raiz = new Nodo(nullptr, e, nullptr);
    }

    bool es_nulo() { return raiz == nullptr; }

    bool es_hoja() {
        if (es_nulo()) return false;
        return raiz->izq == nullptr && raiz->der == nullptr;
    }

    bool es_ult() {
        if (es_nulo()) return false;
        return (raiz->izq == nullptr || raiz->der == nullptr);
    }

    int num_nodos() { return num_nodos(raiz); }
    int num_nodos(Nodo* n) {
        if (!n) return 0;
        return 1 + num_nodos(n->izq) + num_nodos(n->der);
    }

    int altura() { return altura(raiz); }
    int altura(Nodo* n) {
        if (!n) return 0;
        return 1 + max(altura(n->izq), altura(n->der));
    }

    int camino_min() { return camino_min(raiz); }
    int camino_min(Nodo* n) {
        if (!n) return 0;
        return 1 + min(camino_min(n->izq), camino_min(n->der));
    }

    bool es_completo() {
        return altura() == camino_min();
    }

    int num_hojas() { return num_hojas(raiz); }
    int num_hojas(Nodo* n) {
        if (!n) return 0;
        if (!n->izq && !n->der) return 1;
        return num_hojas(n->izq) + num_hojas(n->der);
    }

    void imprimir_preorden() { imprimir_preorden(raiz); cout << endl; }
    void imprimir_preorden(Nodo* n) {
        if (!n) { cout << "null "; return; }
        cout << n->elem << " ";
        imprimir_preorden(n->izq);
        imprimir_preorden(n->der);
    }
};

int main() {
    // Construcción del árbol:
    //       3
    //      / \
    //     8   5
    //    /   / \
    //   2   1   6
    //      /
    //     9

    ArbolBin<int>* nulo = new ArbolBin<int>();

    ArbolBin<int>* nodo2 = new ArbolBin<int>(nulo, 2, nulo);
    ArbolBin<int>* nodo8 = new ArbolBin<int>(nodo2, 8, nulo);

    ArbolBin<int>* nodo9 = new ArbolBin<int>(nulo, 9, nulo);
    ArbolBin<int>* nodo1 = new ArbolBin<int>(nulo, 1, nulo);
    ArbolBin<int>* nodo6 = new ArbolBin<int>(nodo9, 6, nulo);
    ArbolBin<int>* nodo5 = new ArbolBin<int>(nodo1, 5, nodo6);

    ArbolBin<int>* arbol = new ArbolBin<int>(nodo8, 3, nodo5);

    // Demostración
    cout << "=== ARBOL BINARIO ===" << endl;
    cout << "Preorden: ";
    arbol->imprimir_preorden();

    cout << "Es nulo? " << (arbol->es_nulo() ? "Sí" : "No") << endl;
    cout << "Número de nodos: " << arbol->num_nodos() << endl;
    cout << "Número de hojas: " << arbol->num_hojas() << endl;
    cout << "Altura: " << arbol->altura() << endl;
    cout << "Camino mínimo: " << arbol->camino_min() << endl;
    cout << "Es completo? " << (arbol->es_completo() ? "Sí" : "No") << endl;

    return 0;
}
